walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,11 +13,92 @@
 // limitations under the License.
 
 package com.google.sps;
-
+// import com.google.sps.servlets.TimeRange;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.Set;
+import java.util.HashSet;
+import java.util.List;
+import java.util.ArrayList;
+import java.util.Arrays;
 
+/**
+ * Returns possible meeting times on a certain day given a meeting request.
+ */
 public final class FindMeetingQuery {
+
+  public static final int MINUTES_IN_DAY = 1440;
+
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> ret = new ArrayList<TimeRange>();
+    long duration = request.getDuration();
+
+    // no options for too long of a request
+    if (duration > MINUTES_IN_DAY){
+      return ret;
+    }
+
+    Set<String> mandatoryAttendees =  new HashSet<String>();
+    mandatoryAttendees.addAll(request.getAttendees()); 
+
+    Set<String> optionalAttendees = new HashSet<String>();
+    optionalAttendees.addAll(request.getOptionalAttendees());
+
+    Set<String> allAttendees = new HashSet<String>();
+    allAttendees.addAll(mandatoryAttendees);
+    allAttendees.addAll(optionalAttendees);
+
+    // options for no attendees
+    if (allAttendees.size() == 0) {
+      ret.add(TimeRange.WHOLE_DAY);
+      return ret;
+    }
+
+    // find free times
+    List<TimeRange> freeTimes = new ArrayList<TimeRange>(findFreeTimes(allAttendees, events, duration));
+    if (freeTimes.isEmpty()) {
+      // ignore optional attendees if there are no free times across both mandatory and optional attendees
+      freeTimes = new ArrayList<TimeRange>(findFreeTimes(mandatoryAttendees, events, duration));
+    } else {
+      return freeTimes;
+    }
+
+    // adhere to additional test requirement
+    if (mandatoryAttendees.isEmpty()) {
```

Comment:
if mandatoryAttendees are empty, then should we be returning TimeRange.WHOLE_DAY instead? 

also, let's do mandatoryAttendees.isEmpty() check beforefreeTimes = new ArrayList<TimeRange>(findFreeTimes(mandatoryAttendees, events, duration));

right now if mandatoryAttendees is empty, we're making an unnecessary call to findFreeTimes

Reply:
That's what I had at first, but the walkthrough asks to implement a test where if there are no mandatory attendees and no free times for optional attendees, the code returns an empty list